### ISSUE TITLE
SSL server endpoints string syntax results in creation of DefaultOpenSSLContextFactory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6286]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6286|
|Type|enhancement|
|Created|2013-01-29 16:44:40Z|
|Branch|https://github.com/twisted/twisted/tree/sslendpoint-certificateoptions-6286|

`twisted.internet.endpoints._parseSSL` is the function used to parse string syntax for SSL endpoints. It creates a `DefaultOpenSSLContextFactory`. A more useful thing to do would be to create a `twisted.internet.ssl.CertificateOptions` instance, since that is going to be where we concentrate future development (e.g. [#2061](https://github.com/twisted/twisted/issues/2061)).

Attachments:

* [move-_parseSSL-to-CertificateOptions-6286.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/639/6391cdba4f8383004f181d813f7a5625ffbedf23/3fbbd2bb52eb4c2d04cba221333db0a13e7e5dca.patch) (6302 bytes) - added by hynek on 2013-02-03 16:20:15Z - Move to CertificateOptions, push test coverage of contractor.
* [move-_parseSSL-to-CertificateOptions-6286-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/639/6391cdba4f8383004f181d813f7a5625ffbedf23/b50821b910385d838b1e9649146f5e616c115d53.patch) (6511 bytes) - added by hynek on 2013-03-02 17:54:41Z - Same as before, using SSLv23 as default method.
* [move-_parseSSL-to-CertificateOptions-6286-v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/639/6391cdba4f8383004f181d813f7a5625ffbedf23/1ae0acd22b865c6f0f8ceacea39b7a0c152bb78f.patch) (6832 bytes) - added by hynek on 2013-03-02 20:24:59Z - Brownbag fix of skip definition.
* [fix-concat-order-6286.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/639/6391cdba4f8383004f181d813f7a5625ffbedf23/37981826c32712fdf02817c7a865fc69ee8e1242.patch) (2299 bytes) - added by hynek on 2013-03-07 14:27:47Z - Use a more natural concatenation order.
* [untangle-6286.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/639/6391cdba4f8383004f181d813f7a5625ffbedf23/4391c642d35b5ffa3eb65b7719398c3b230d5b9c.patch) (3080 bytes) - added by hynek on 2013-03-08 17:42:12Z - Untangled as per jp’s request.

<details><summary>Searchable metadata</summary>

```
trac-id__6286 6286
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_sslendpoint_certificateoptions_6286 branches/sslendpoint-certificateoptions-6286
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1359477880000000 1359477880000000
changetime__1362771980000000 1362771980000000
version__None None
owner__tom_prince tom.prince
cc__hynek cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

(In [37302]) Branching to 'sslendpoint-certificateoptions-6286'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Note that IIRC current ssl endpoint defaults to SSLv23 (backwards compat TLS1+SSL3) with SSL2 disabled. New code in patch defaults to TLSv1 method. So that should be fixed if my memory is correct.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_7"></a>|@hynek commented|
|-|-|

Itamar is right.

But since TLSv1 is from 1999…what would it take to make it the (more secure…) default for such an high-level function?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj set owner to @hynek<br>@wallrj set status to `new`|
|-|-|

Hynek,

Here's a quick (inexpert) review of your patch.

 1. Patch applied cleanly to a new branch
    (source:branches/sslendpoint-certificateoptions-6286)
 2. Missing news file (I added one to the branch in r37304)
 3. [Build results look fine](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sslendpoint-certificateoptions-6286)
 4. See Itamar's comment (above)
    1. He's right, the old default was
       [https://twistedmatrix.com/documents/current/api/twisted.internet.ssl.DefaultOpenSSLContextFactory.html#__init__ sslmethod=SSL.SSLv23_METHOD]
    2. New default will be
       [https://twistedmatrix.com/documents/current/api/twisted.internet.ssl.CertificateOptions.html#__init__ TLSv1_METHOD]
    3. So that needs changing for backwards compatibility
    4. And maybe a test to demonstrate that SSL.SSLv23_METHOD is the
       desired default.

Please address point 4 and resubmit for review.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_9"></a>|@hynek commented|
|-|-|

Thank you!

FTR: We need to fix [#6337](https://github.com/twisted/twisted/issues/6337) before we can use `SSLv23` by default.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)|@hynek removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

The patch has a bug in the skip definition.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_12"></a>|@hynek commented|
|-|-|

Thanks JP, fixed!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

As a note, when attaching a patch against a ticket that has a branch, please make the patch against the branch, rather than trunk. (I've dealt with it here already)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_14"></a>|@tomprince set owner to zzccflpccl|
|-|-|

The build results look good. The code change look correct.

This code looks like it enables `SSL.OP_SINGLE_DH_USE` and sets a session-id, which `DefaultOpenSSLContextFactory` doesn't. I suspect neither of these changes is going to be an issue, but I'm not sure.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @hynek|
|-|-|

I'm not worried about either of those changes.  They're minimally disruptive (ie, neither of them will make connections stop succeeding or compromise any of the security properties of those connections).  Applications that require such precise control over SSL parameters are surely already using their own context factories anyway.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

```
<exarkun> I only glanced at the actual code changes.  If you're happy with those, I'm happy with the SSL changes you pointed out would be made.
<tomprince> k.
<tomprince> I *think* those are only differences between the two factories. (and the verify depth, but the CertificatOptions docstring claims it sets it to the default)
<exarkun> Hum
<exarkun> CertificateOptions also disables session tickets by default (increases interoperability)
<exarkun> Uh
<exarkun> Oh
<exarkun> and CertificateOptions has lots of confusing verification logic
<exarkun> requireCertificate is True
<exarkun> but verify is False
<exarkun> and caCerts is None
<exarkun> okay, which means it doesn't require a cert, which is the same
<exarkun> uhh cert chains, why are you such a mess
<exarkun> eh I'm pretty sure the DefaultOpenSSLContextFactory use_certificate_file call has the same meaning as the CertificateOptions use_certificate call
<exarkun> does loadPEM work with two PEMs smashed together?
<exarkun> yea looks like
<exarkun> so probably fine
<exarkun> although looking at the code
<exarkun> PrivateCertificate.load(certPEM + keyPEM).options(...) is probably a somewhat nicer spelling
```

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_18"></a>|@hynek commented|
|-|-|

I’m not entirely sure how to proceed here since the review tag has been removed without further actual criticism. Is it ready to merge or should I change something?

I’m not sure what to make of

    <exarkun> PrivateCertificate.load(certPEM + keyPEM).options(...) is probably a somewhat nicer spelling

Is it about the order of the concatenation? If so, I’ll attach a quick fix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

> Is it about the order of the concatenation? If so, I’ll attach a quick fix.

No, the order doesn't matter.  It's about using `PrivateCertificate.options` instead of `CertificateOptions`.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_20"></a>|@hynek commented|
|-|-|

As per discussion on IRC, the approach with `PrivateCertificate.options` doesn’t work with the current API.

Since nothing is left open, merging would be appreciated.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)|@hynek removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @tomprince|
|-|-|

This code is correct:

```python
    cf = ssl.CertificateOptions(
        privateKey=ssl.PrivateCertificate.loadPEM(
            keyPEM + certPEM).privateKey.original,
        certificate=ssl.Certificate.loadPEM(certPEM).original, **kw)
```

but quite a mess.  Please clean it up as much as possible (eg, split it into multiple statements and by not loading `certPEM` twice).  Also file a ticket about how `PrivateCertificate.options` is greatly hampered in its utility by the lack of support for any options other than certificate authorities.

Otherwise, fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_24"></a>|@hynek commented|
|-|-|

Changes done, ticket filed.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_25"></a>|@tomprince set status to `closed`|
|-|-|

(In [37463]) Merge sslendpoint-certificateoptions-6286: Use twisted.internet.ssl.CertificateOptions in strports.

Author: hynek
Reviewers: rwall, tom.prince, exarkun
Fixes: [#6286](https://github.com/twisted/twisted/issues/6286)
Refs: [#2601](https://github.com/twisted/twisted/issues/2601), [#6377](https://github.com/twisted/twisted/issues/6377)

twisted.internet.endpoints._parseSSL is the function used to
parse string syntax for SSL endpoints. It creates a
DefaultOpenSSLContextFactory. A more useful thing to do
would be to create a twisted.internet.ssl.CertificateOptions
instance, since that is going to be where we concentrate
future development (e.g. [#2061](https://github.com/twisted/twisted/issues/2061)).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

This looks good.
